Exit Transformer cURL examples in demo section; remove obsolete handle_admin field; update for errs by mimetype (#2332)
Kong/docs.konghq.com: app/_hub/kong-inc/exit-transformer/index.md

desc: Customize Kong exit responses sent downstream
description: |
  Transform and customize Kong response exit messages using Lua functions.

>   The capabilities range from changing messages, status codes, and headers

Reviewer: ```suggestion
  The capabilities range from changing messages, status codes, and headers,
```

    - name: handle_unknown
      default: "`false`"
      required: false
      description: Allow transform to apply to unmatched Service, Route, or Workspace (404) responses.
    - name: handle_unexpected
      default: "`false`"
      required: false
      description: Allow transform to apply to unexpected request (400) responses.

---

## Transforming 4xx and 5xx Responses

By default, the Exit Transformer is only applied to requests that match its
criteria (its Route or Service matching configuration) or globally within a Workspace.

### Handling Unmatched 400 and 404 Responses

Requests that result in 400 or 404 responses neither match any criteria nor fall
within any specific Workspace, and standard plugin criteria will never match those
responses. You can designate Exit Transformer configurations that _do_ handle these
responses by enabling the `handle_unexpected` (400) and `handle_unknown` (404) settings.
The `handle_unknown` parameter should only be enabled on a single plugin configuration.

> The `handle_unexpected` parameter can be enabled on as many plugin configurations

Reviewer: Suggest breaking these two lines into bullet points for better readability, and to break up this block of text.

Author: Done

    - name: handle_unknown
      default: "`false`"
      required: false
      description: Allow transform to apply to unmatched Service, Route, or Workspace

>       (404) responses. This flag only applies to one configuration entry.

Author: Have requested clarification from Lluis on whether the flag applies to both unknown and expected because there is a discrepancy per the readme. Also, the template has issues with line 41 column 7 no matter how I format the additional information. Will add to a note section under params table after Lluis confirms.

end
```

> #### Modify the body and headers regardless if provided {#mod-body-head}

Reviewer: "Regardless if" is an awkward construction, not sure that's grammatically correct. I believe the proper construction is "regardless of if", which is even more awkward in this case - I'd avoid using "regardless" entirely to simplify. 

Struggling to come up with a good phrase here though. I actually prefer the original phrasing over the rephrase, but that might be a personal choice.

Author: stet

## Demonstration

### Step 1: Create a Service in Kong

   {% navtabs %}
   {% navtab Using cURL %}

   ```bash

>    curl -i -X POST http://<admin-hostname>:8001/services \

Reviewer: ```suggestion
   $ curl -i -X POST http://<admin-hostname>:8001/services \
```

    --data name=example.com \
    --data url='http://mockbin.org'
   ```

   {% endnavtab %}
   {% navtab Using HTTPie %}

   ```bash
   $ http :8001/services name=example.com host=mockbin.org
   ```

   {% endnavtab %}
   {% endnavtabs %}

### Step 2: Create a Route in Kong

   {% navtabs %}
   {% navtab Using cURL %}

```bash

> curl -i -X POST http://<admin-hostname>:8001/services/example.com/routes \

Reviewer: ```suggestion
$ curl -i -X POST http://<admin-hostname>:8001/services/example.com/routes \
```

## Demonstration

### Step 1: Create a Service in Kong

   {% navtabs %}
   {% navtab Using cURL %}

   ```bash
   curl -i -X POST http://<admin-hostname>:8001/services \
    --data name=example.com \
    --data url='http://mockbin.org'
   ```

   {% endnavtab %}
   {% navtab Using HTTPie %}

   ```bash
   $ http :8001/services name=example.com host=mockbin.org
   ```

   {% endnavtab %}
   {% endnavtabs %}

### Step 2: Create a Route in Kong

   {% navtabs %}
   {% navtab Using cURL %}

```bash
curl -i -X POST http://<admin-hostname>:8001/services/example.com/routes \

>   --data 'paths[]=/mock' \
>   --data name=mocking

Reviewer 2: To align this example with the other steps or the httpie command, it should add the route as a host and not a path

```suggestion
  --data 'hosts[]=example.com'
```

>       return status, body, headers

Reviewer 2: My bad on this one (I think?). It should read `return status, new_body, headers`

>       return status, body, headers

Reviewer 2: ```suggestion
      return status, new_body, headers
```

    end
```

### Step 4: Configure the Plugin with its Transform

Configure the `exit-transformer` plugin with `transform.lua`.

   {% navtabs %}
   {% navtab Using cURL %}

   ```bash
   $ curl -X POST http://<admin-hostname>:8001/services/example.com/plugins \

>     --data "name=exit-transformer"  \
>     --data "config.functions=@transform.lua"

Reviewer 2: ```suggestion
    -F "name=exit-transformer"  \
    -F "config.functions=@transform.lua"
```

        "status": 401,
        "kong_message": "No API key found in request, arr!"
    }
```

## More Examples

### Apply the Plugin Globally to Handle Unknown Responses

> Note the plugin can also be applied globally:

Reviewer: ```suggestion
The plugin can also be applied globally:
```

{% navtabs %}
{% navtab Using cURL %}

```bash

> curl -X POST http://<admin-hostname>:8001/plugins/ \

Reviewer: ```suggestion
$ curl -X POST http://<admin-hostname>:8001/plugins/ \
```

    -F "name=exit-transformer"  \
    -F "config.handle_unknown=true" \
    -F "config.functions=@transform.lua"
...

> curl --header 'Host: non-existent.com' 'localhost:8000'

Reviewer: ```suggestion
$ curl --header 'Host: non-existent.com' 'localhost:8000'
```

```

{% endnavtab %}
{% navtab Using HTTPie %}

```bash
$ http :8001/plugins \
    name=exit-transformer \
    config.handle_unknown=true \

>     config.functions=@transform.lua
> 
>     $ http :8000 Host:non-existent.com

Reviewer: ```suggestion
    config.functions=@transform.lua

$ http :8000 Host:non-existent.com
```

```

{% endnavtab %}
{% endnavtabs %}

Response:

```bash  
    HTTP/1.1 200 OK
    ...
    X-Some-Header: some value

    {
        "error": true,
        "status": 404,
        "kong_message": "No Route matched with those values, arr!"
    }
  ```

### Custom Errors by MIME Type

This example shows a use case where you want custom JSON and HTML responses
based on an [Accept header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept).

> Create a file named `custom-errors-by-mimetype.lua` file with the transformation

Reviewer: ```suggestion
Create a file named `custom-errors-by-mimetype.lua` with the transformation
```

> code shown below. See the full list of HTTP response codes [above](#http-msgs).
> Include the status codes you want to customize. Any status code not listed in the

Reviewer: ```suggestion
code shown below. Based on the list of HTTP response codes [above](#http-msgs),
include the status codes you want to customize. Any status code not listed in the
```
Either this option, or flip the two sentences (Include the status codes you want to customize. See the full list...)

Author: Will flip because "Based on ..."  seems backwards

  end,
}

return function(status, body, headers)
  if status < 400 then
    return status, body, headers
  end

  local accept = kong.request.get_header("accept")
  -- Gets just first accept value. Can be improved to be compliant quality
  -- etc parser. Look into kong.pdk.response get_response_type
  if type(accept) == "table" then
    accept = accept[1]
  end
  accept = split(accept, ",")[1]

  if not formats[accept] then
    return status, body, headers
  end

  return formats[accept](status, get_message(status), headers)
end
```

Configure the `exit-transformer` plugin with `custom-errors-by-mimetype.lua`.

{% navtabs %}
{% navtab Using cURL %}

```bash
$ curl -X POST http://<admin-hostname>:8001/services/example.com/plugins \
 -F "name=exit-transformer"  \
 -F "config.handle_unknown=true" \
 -F "config.handle_unexpected=true" \
 -F "config.functions=@examples/custom-errors-by-mimetype.lua"
```

{% endnavtab %}
{% navtab Using HTTPie %}

```bash
$ http -f :8001/plugins name=exit-transformer \
  config.handle_unknown=true \
  config.handle_unexpected=true \
  config.functions=@examples/custom-errors-by-mimetype.lua

> ```

Reviewer: Comment applies to entire doc: be careful with the indents on code examples. It won't make a difference in the end for whether the command works or not, but for visual consistency+clean copyable code, it's nice to have them follow the same style. As as rule, I've been trying to line things up the way you have for the HTTPie example here: align each line with the start of the command, not the $.

Author: > Comment applies to entire doc: be careful with the indents on code examples. It won't make a difference in the end for whether the command works or not, but for visual consistency+clean copyable code, it's nice to have them follow the same style. As as rule, I've been trying to line things up the way you have for the HTTPie example here: align each line with the start of the command, not the $.

@lena-larionova i see what you mean. I just committed reviewer comments in a batch and didn't really notice the difference. I need new glasses.

Author: @lena-larionova  I did just notice the discrepancy with the autogenerated examples though. Is there any way we can get those to line up as you suggested for the non-generated examples?

![image](https://user-images.githubusercontent.com/3756245/95138222-5d978300-072f-11eb-9eed-661cf0115996.png)

Author: @lena-larionova this PR is ready for another review.

        "status": 401,
        "kong_message": "No API key found in request, arr!"
    }
```

## More Examples

### Apply the Plugin Globally to Handle Unknown Responses

Note the plugin can also be applied globally:

{% navtabs %}
{% navtab Using cURL %}

```bash
curl -X POST http://<admin-hostname>:8001/plugins/ \

>     --data "name=exit-transformer"  \
>     --data "config.handle_unknown=true" \
>     --data "config.functions=@transform.lua"

Reviewer 2: ```suggestion
    -F "name=exit-transformer"  \
    -F "config.handle_unknown=true" \
    -F "config.functions=@transform.lua"
```

...
curl --header 'Host: non-existent.com' 'localhost:8000'
```

{% endnavtab %}
{% navtab Using HTTPie %}

```bash
$ http :8001/plugins \
    name=exit-transformer \
    config.handle_unknown=true \
    config.functions=@transform.lua

    $ http :8000 Host:non-existent.com
```

{% endnavtab %}
{% endnavtabs %}

Response:

```bash  
    HTTP/1.1 200 OK
    ...
    X-Some-Header: some value

    {
        "error": true,
        "status": 404,
        "kong_message": "No Route matched with those values, arr!"
    }
  ```

### Custom Errors by MIME Type

This example shows a use case where you want custom JSON and HTML responses
based on an [Accept header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept).

> 

Reviewer 2: We still need to include the code of `custom-errors-by-mimetype.lua` right? https://github.com/Kong/kong-plugin-enterprise-exit-transformer/blob/master/examples/custom-errors-by-mimetype.lua

It's a tad long though

Author: > We still need to include the code of `custom-errors-by-mimetype.lua` right? https://github.com/Kong/kong-plugin-enterprise-exit-transformer/blob/master/examples/custom-errors-by-mimetype.lua
> 
> It's a tad long though

@eskerda yes it is otherwise it would already be there. Since the response codes are now included, I could just have a comment line that says see codes section above?

Reviewer 2: But then the code sample will not work.. maybe we can add it as:

```
local HTTP_MESSAGES = {
    s400 = "Bad request",
    s401 = "Unauthorized",
    -- ...
    -- See codes section above for the full list
    s511 = "Network authentication required",
    default = "The upstream server responded with %d"
}
```

or similar. Otherwise, we can always put the code sample on a github gist, or a different page and link to it?

Author: Added abbreviate code example as we discussed.

{% navtabs %}
{% navtab Using cURL %}

```bash
$ curl -X POST http://<admin-hostname>:8001/services/example.com/plugins \

>  --data "name=exit-transformer"  \
>  --data "config.handle_unknown=true" \
>  --data "config.handle_unexpected=true" \
>  --data "config.functions=@examples/custom-errors-by-mimetype.lua"

Reviewer 2: ```suggestion
 -F "name=exit-transformer"  \
 -F "config.handle_unknown=true" \
 -F "config.handle_unexpected=true" \
 -F "config.functions=@examples/custom-errors-by-mimetype.lua"
```
